### PR TITLE
What's Here for Kernel

### PR DESCRIPTION
**Pull request by BurdetteLamar**
These sections may not be in the best order.

I'll add the TOC when everything else is settled.

### CODE REVIEW

File: object.c [hunk 1]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *           +at_exit+ handlers.
+     * - #exit!:: Exits the current process without calling any registered
+     *            +at_exit+ handlers.
+     *
+     * === Exceptions
+     *
+     * - #catch:: Executes the given block, possibly catching a thrown object.
+     * - #fail:: Raises an exception based on the given arguments.
+     * - #raise:: Raises an exception based on the given arguments.
+     * - #throw:: Transfers control to the end of the active catch block
```
Comment (jeremyevans) [inline object.c:4626]: `Returns from the active catch block waiting for the given tag.`

File: object.c [hunk 2]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *            that is waiting for the given tag.
+     *
+     * === Strings
+     *
+     * - #format:: Returns the string resulting from applying the given format string
```
Comment (jeremyevans) [inline object.c:4631]: `format` is an alias of `sprintf`.

File: object.c [hunk 3]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *             to any additional arguments.
+     * - #gsub:: Performs global substitution using the given pattern and either
```
Comment (jeremyevans) [inline object.c:4633]: ... `given pattern and replacement.`  Same for `sub`.

File: object.c [hunk 4]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     * - #sub:: Performs single substitution using the given pattern and either
+     *          a replacement string or a block.
+     *
+     * === \IO
+     *
+     * - #chomp:: Removes the trailing record separator, if any, from <tt>$_</tt>.
+     * - #chop:: Removes the last character, if any, from <tt>$_</tt>.
+     * - #gets:: Returns and assigns to <tt>$_</tt> the next line from the current input.
+     * - #open:: Creates an IO object connected to the given stream, file, or subprocess.
+     * - #p::  Writes the given objects to standard output.
```
Comment (jeremyevans) [inline object.c:4646]: `Prints the given objects inspect output to the standard output`

File: object.c [hunk 5]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     * - #pp:: Prints the given objects in pretty form.
+     * - #print:: Prints the given objects to standard output.
```
Comment (jeremyevans) [inline object.c:4648]: ... `without a newline.`

File: object.c [hunk 6]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     * - #printf:: Prints the string resulting from applying the given format string
+     *             to any additional arguments.
+     * - #putc:: Equivalent to <tt.$stdout.putc(object)</tt> for the given object.
+     * - #puts:: Equivalent to <tt>$stdout.puts(*objects)</tt> for the given objects.
+     * - #readline:: Equivalent to Kernel.gets, except that +readline+
```
Comment (jeremyevans) [inline object.c:4653]: `Similar to gets, but raises an exception at the end of file.`

File: object.c [hunk 7]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *               raises an exception at end of file.
+     * - #readlines:: Returns an array of the remaining lines from the current input.
+     * - #select:: From arrays of IO streams, returns the first one to become ready.
+     *
+     * === Procs
+     *
+     * - #lambda:: Returns a new Proc formed from the given block.
```
Comment (jeremyevans) [inline object.c:4660]: `Returns a lambda proc for the given block.`

File: object.c [hunk 8]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *           a replacement string or a block.
+     * - #sprintf:: Returns the string resulting from applying the given format string
+     *              to any additional arguments.
+     * - #sub:: Performs single substitution using the given pattern and either
+     *          a replacement string or a block.
+     *
+     * === \IO
+     *
+     * - #chomp:: Removes the trailing record separator, if any, from <tt>$_</tt>.
+     * - #chop:: Removes the last character, if any, from <tt>$_</tt>.
```
Comment (jeremyevans) [inline object.c:4643]: Both `chomp` and `chop` are not defined by default, and only defined if the `-p` flag is given.  I don't think either should be documented.

File: object.c [hunk 9]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     * - #pp:: Prints the given objects in pretty form.
+     * - #print:: Prints the given objects to standard output.
+     * - #printf:: Prints the string resulting from applying the given format string
+     *             to any additional arguments.
+     * - #putc:: Equivalent to <tt.$stdout.putc(object)</tt> for the given object.
+     * - #puts:: Equivalent to <tt>$stdout.puts(*objects)</tt> for the given objects.
+     * - #readline:: Equivalent to Kernel.gets, except that +readline+
+     *               raises an exception at end of file.
+     * - #readlines:: Returns an array of the remaining lines from the current input.
+     * - #select:: From arrays of IO streams, returns the first one to become ready.
```
Comment (jeremyevans) [inline object.c:4656]: `Same as <tt>IO.select</tt>.`

File: object.c [hunk 10]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     * === Tracing
+     *
+     * - #set_trace_func:: Sets the given proc as the handler for tracing,
+     *                     or disables tracing if given +nil+.
+     * - #trace_var:: Specifies the tracing of assignments to the given global variable.
+     * - #untrace_var:: Disables tracing of assignments to the given global variable.
+     *
+     * === Subprocesses
+     *
+     * - #fork:: Creates a subprocess that executes the given block, if any.
```
Comment (jeremyevans) [inline object.c:4672]: `Forks the current process into two processes.`

File: object.c [hunk 11]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     * - #spawn:: Executes the given command and returns its pid without waiting.
```
Comment (jeremyevans) [inline object.c:4673]: ... `for completion.`

File: object.c [hunk 12]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *               is first referenced.
+     * - #load:: Loads and executes the Ruby program found in the given file.
+     * - #require:: Loads the given file.
+     * - #require_relative:: Loads the given relative file.
+     *
+     * === Yielding
+     *
+     * - #tap:: Yields +self+ to the given block; returns +self+.
+     * - #then:: Yields +self+ to the block and returns the result of the block.
+     * - #yield_self:: Yields +self+ to the block and returns the result of the block.
```
Comment (jeremyevans) [inline object.c:4687]: `yield_self` is an alias of `then`.

File: object.c [hunk 13]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *
+     * === \Random Values
+     *
+     * - #rand:: Returns a pseudo-random floating point number
+     *           strictly between 0.0 and 1.0.
+     * - #srand:: Seeds the system pseudo-random number generator with the given number.
+     *
+     * === Other
+     *
+     * - #callcc:: Generates a Continuation object, which it passes
```
Comment (jeremyevans) [inline object.c:4697]: This isn't defined by default (you need to require `continuation`), so it shouldn't be documented.

File: object.c [hunk 14]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *
+     * - #tap:: Yields +self+ to the given block; returns +self+.
+     * - #then:: Yields +self+ to the block and returns the result of the block.
+     * - #yield_self:: Yields +self+ to the block and returns the result of the block.
+     *
+     * === \Random Values
+     *
+     * - #rand:: Returns a pseudo-random floating point number
+     *           strictly between 0.0 and 1.0.
+     * - #srand:: Seeds the system pseudo-random number generator with the given number.
```
Comment (jeremyevans) [inline object.c:4693]: `Seeds the system` -> `Seeds the`

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Thanks for working on this.  See inline recommendations (31 of them).

**Comment by BurdetteLamar** [comment]
Re: chomp, chop, dup, clone, callcc.   I took my cue from https://ruby-doc.org/core-3.0.1/Kernel.html.  Is that not correct?

**Comment by jeremyevans** [comment]
> Re: chomp, chop, dup, clone, callcc. I took my cue from https://ruby-doc.org/core-3.0.1/Kernel.html. Is that not correct?

You cannot take RDoc output as gospel. The Kernel `sub`, `gsub`, `chop`, and `chomp` methods are only defined if ruby is run in loop mode (either `-p` or `-n` flags): https://github.com/ruby/ruby/blob/master/ruby.c#L1998

RDoc puts `dup` in the Object documentation even though it is defined in Kernel.  This isn't too bad, as that's where it documents other Kernel methods you expect to call with an explicit receiver.  However, because `clone` is implemented in Ruby and not C, it doesn't use the same logic to move the method documentation to Object, which is why `clone` now shows up in the Kernel documentation.  Personally, I think all Kernel methods should be documented in Kernel, but I don't make the rules. :)

`callcc` shows up because the code that RDoc parses is in a C file in the root directory.  However, it's inside a C function that is not called unless you call `require 'continuation'`.  So a similar reason to why `sub`, `gsub`, `chop`, and `chomp`  show up.  RDoc is just scanning the C file to look for something that appears to be a method definition, it cannot reasonably have an understanding of the surrounding context.

File: object.c (hunk 9 above)
Comment (BurdetteLamar) [reply object.c]: I've let this link through, instead of tt.

File: object.c [hunk 15]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *
+     * === Other
+     *
+     * - #callcc:: Generates a Continuation object, which it passes
+     *             to the associated block.
+     * - #clone:: Returns a shallow copy of +self+.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply object.c]: I've omitted #clone here, will do both over at What's Here in Object.

**Comment by BurdetteLamar** [comment]
@jeremyevans, I know there would be a lot for you to do here, so thanks much for doing it.

File: object.c [hunk 16]
```diff
@@ -4565,6 +4565,143 @@ InitVM_Object(void)
...
+     *                   in the current context.
+     * - #caller:: Returns the current execution stack as an array of strings.
+     * - #caller_locations:: Returns the current execution stack as an array of strings.
+     * - #class:: Returns the class of +self+.
+     * - #frozen?:: Returns whether +self+ is frozen.
+     * - #global_variables:: Returns an array of global variables as symbols.
+     * - #local_variables:: Returns an array of local variables as symbols.
+     * - #test:: Performs specified tests on the given single file or pair of files.
+     *
+     * === Executing
```
[the remark this answers is not in the document]
Comment (jeremyevans) [reply object.c:4618]: I would move `eval` and `syscall` to the Other section.

File: object.c (hunk 2 above)
Comment (jeremyevans) [reply object.c:4640]: You updated the `sprintf` line to mention the `format` alias, so I would remove the `format` line

File: object.c [hunk 17]
```diff
@@ -4565,6 +4565,151 @@ InitVM_Object(void)
...
+     *
+     * - #catch:: Executes the given block, possibly catching a thrown object.
+     * - #raise (aliased as #fail):: Raises an exception based on the given arguments.
+     * - #throw:: Returns from the active catch block waiting for the given tag.
+     *
+     * === Strings
+     *
+     * - #format:: Returns the string resulting from applying the given format string
+     *             to any additional arguments.
+     * - #gsub:: Performs global substitution using the given pattern and replacement'
```
Comment (jeremyevans) [inline object.c:4642]: `'` at end of line should be `.`

File: object.c [hunk 18]
```diff
@@ -4565,6 +4565,151 @@ InitVM_Object(void)
...
+     * - #sprintf (aliased as #format):: Returns the string resulting from applying
+     *                                   the given format string
+     *                                   to any additional arguments.
+     * - #sub:: Performs single substitution using the given given pattern
```
Comment (jeremyevans) [inline object.c:4646]: `given given` -> `given`

File: object.c [hunk 19]
```diff
@@ -4565,6 +4565,151 @@ InitVM_Object(void)
...
+     *          and replacement.
+     *
+     * === \IO
+     *
+     * - #gets:: Returns and assigns to <tt>$_</tt> the next line from the current input.
+     * - #open:: Creates an IO object connected to the given stream, file, or subprocess.
+     * - #p::  Prints the given objects inspect output to the standard output.
```
Comment (jeremyevans) [inline object.c:4653]: `objects inspect` -> `objects' inspect`.  Sorry, I realize I provided the wrong text. :)

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks good, only a few suggestions.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Doesn't look like any of the inline suggestions is addressed.  Missed push?

**Comment by BurdetteLamar** [comment]
> 
> 
> Doesn't look like any of the inline suggestions is addressed. Missed push?

Yes.  Now pushed.

File: object.c [hunk 20]
```diff
@@ -4565,6 +4565,146 @@ InitVM_Object(void)
      *
      *   sprintf "%.1f", 1.234 #=> "1.2"
      *
+     * == What's Here
+     *
+     * \Module \Kernel provides methods that are useful for:
+     *
+     * - {Converting}[#module-Kernel-label-Converting]
+     * - {Querying}[#module-Kernel-label-Querying]
+     * - {Executing}[#module-Kernel-label-Executing]
```
Comment (jeremyevans) [inline object.c:4574]: Executing heading was removed, so this line can be deleted.

File: object.c [hunk 21]
```diff
@@ -4565,6 +4565,146 @@ InitVM_Object(void)
...
+     * - #exit!:: Exits the current process without calling any registered
+     *            +at_exit+ handlers.
+     *
+     * === Exceptions
+     *
+     * - #catch:: Executes the given block, possibly catching a thrown object.
+     * - #raise (aliased as #fail):: Raises an exception based on the given arguments.
+     * - #throw:: Returns from the active catch block waiting for the given tag.
+     *
+     * === Strings
```
Comment (jeremyevans) [inline object.c:4633]: `gsub` and `sub` are only defined in loop mode and should not be listed.  I should have noted this earlier, sorry.  I would move `sprintf` to other, and delete the Strings heading from the TOC.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Almost ready.  Only a couple of suggestions now.

**Review by jeremyevans — APPROVED** [review]
Looks good, thanks!

**Pull request merged**